src/metabase/db/metadata_queries.clj

```diff
@@ -36,39 +36,47 @@
-        ;; Ingestion time partitioned table can use either _PARTITIONDATE or _PARTITIONTIME as
-        ;; partitioned field
-        partition-field (or (t2/select-one :model/Field
-                                           :table_id table-id
-                                           :database_partitioned true
-                                           :active true
-                                           ;; prefer _PARTITIONDATE over _PARTITIONTIME for ingestion time query
-                                           {:order-by [[:name :asc]]})
-                            (throw (ex-info (format "No partitioned field found for table: %d" table-id)
-                                            {:table_id table-id})))
-        filter-form     (partition-field->filter-form partition-field)]
-    (update query :filter (fn [existing-filter]
-                            (if (some? existing-filter)
-                              [:and existing-filter filter-form]
-                              filter-form)))))
+(defn add-required-filter-if-needed
+  "Add a dummy filter for tables that require a filters.
+  Look into tables from source tables and all the joins.
+  Currently this only apply to partitioned tables on bigquery that requires a partition filter."
+  [query]
+  (let [table-ids              (->> (conj (keep :source-table (:joins query)) (:source-table query))
+                                    (filter pos-int?))
+        required-filter-fields (when (seq table-ids)
+                                 (t2/select :model/Field {:select    [:f.*]
+                                                          :from      [[:metabase_field :f]]
+                                                          :left-join [[:metabase_table :t] [:= :t.id :f.table_id]]
+                                                          :where     [:and
+                                                                      [:= :f.active true]
+                                                                      [:= :f.database_partitioned true]
+                                                                      [:= :t.active true]
+                                                                      [:= :t.database_require_filter true]
+                                                                      [:in :t.id table-ids]]}))
+        update-query-filter-fn (fn [existing-filter new-filter]
+                                (if (some? existing-filter)
+                                  [:and existing-filter new-filter]
+                                  new-filter))]
+    (case (count required-filter-fields)
+      0
+      query
+      1
+      (update query :filter update-query-filter-fn (partition-field->filter-form (first required-filter-fields)))
+      ;; > 1
+      (update query :filter update-query-filter-fn (into [:and] (map partition-field->filter-form required-filter-fields))))))
```

Comment:
hmm this could be simpler if you modified `update-query-filter-fn` to take a list, like this
```suggestion
      (update query :filter (fn [existing-filter new-filters]
                              (-> [:and] (into new-filters) (into existing-filter)))
        (map partition-field->filter-form required-filter-fields)))))
```
and then you could use `(cond-> query (seq required-filter-fields) (update ...))` instead of the `case` expression

Reply:
that's what I did at first. I changed to `case` because I don't want to add a `:and` if there is only one filter.

Reply:
fair, both approaches have their downsides. If you really care about the extra `:and` I guess you could use this function to remove it, but then we're talking very marginal gains in having less code

```
(defn remove-extra-and [[_and & args :as clause]]
  (if (= 1 (count args))) args clause))
```